Source/ProgressionSystemRuntime/Private/Widgets/PSOverlayWidget.cpp

```diff
@@ -4,11 +4,113 @@
 #include "Widgets/PSOverlayWidget.h"
 
 #include "Components/Image.h"
+#include "Curves/CurveFloat.h"
+#include "Data/PSDataAsset.h"
 
 #include UE_INLINE_GENERATED_CPP_BY_NAME(PSOverlayWidget)
 
-// Sets the visibility of the background overlay and lock icon.
-void UPSOverlayWidget::SetOverlayVisibility(ESlateVisibility VisibilitySlate)
+// Sets the visibility of the overlay elements and playing fade animation if needed
+void UPSOverlayWidget::SetOverlayVisibility(ESlateVisibility VisibilitySlate, bool bShouldPlayFadeAnimation/* = false*/)
+{
+	if (!bShouldPlayFadeAnimation)
+	{
+		SetOverlayItemsVisibility(VisibilitySlate);
+		return;
+	}
+
+	bShouldPlayFadeAnimationInternal = bShouldPlayFadeAnimation;
+
+	if (VisibilitySlate == ESlateVisibility::Visible)
+	{
+		ESlateVisibility PrevOverlayOpacity = PSCBackgroundOverlay->GetVisibility();
+		ESlateVisibility PrevIconOpacity = PSCBackgroundIconLock->GetVisibility();
+
+		//if new visibility is same as previous animation is not required 
+		if (PrevOverlayOpacity == VisibilitySlate && PrevIconOpacity == VisibilitySlate)
+		{
+			bShouldPlayFadeAnimationInternal = false;
+		}
+		bIsFadeInAnimationInternal = true;
+		SetOverlayItemsVisibility(VisibilitySlate);
+	}
+	else
+	{
+		bIsFadeInAnimationInternal = false;
+	}
+
+	const UWorld* World = GetWorld();
+	if (!ensureMsgf(World, TEXT("ASSERT: [%i] %s:\n'World' is not valid!"), __LINE__, *FString(__FUNCTION__)))
+	{
+		return;
+	}
+
+	StartTimeFadeAnimationInternal = World->GetTimeSeconds();
+}
+
+// overrides NativeTick to make the user widget tickable
+void UPSOverlayWidget::NativeTick(const FGeometry& MyGeometry, float InDeltaTime)
+{
+	Super::NativeTick(MyGeometry, InDeltaTime);
+	if (bShouldPlayFadeAnimationInternal)
+	{
+		const bool bIsFinished = !FadeOverlayElementsAnimation(StartTimeFadeAnimationInternal);
```

Comment:
Your `FadeOverlayElementsAnimation` function looks like a helper now. But you don't need it to be helper since it's used only once. Please refactor `FadeOverlayElementsAnimation` function a bit to make it look more clear:

- rename it with Tick prefix to show others is called from tick: `TickPlayFadeAnimation()`
- Remove `ref` argument, you don't need it, just directly modify the `StartTimeFadeAnimationInternal` property inside the TickPlayFadeAnimation function
- you don't need `bool` return, just `void` - move all logic from NativeTick inside the function
- add BlueprintCallable UFUNCTION for content developers

### Expected Result
```cpp
void UPSOverlayWidget::NativeTick(const FGeometry& MyGeometry, float InDeltaTime)
{
       TickPlayFadeAnimation();
}
```

Reply:
done as part of :https://github.com/h4rdmol/ProgressionSystem/pull/20/commits/f112d52784a7a36ec46d4d44980ead28018688e8

Reply:
@h4rdmol next is missing:
> add BlueprintCallable UFUNCTION for content developers

Reply:
added with: https://github.com/h4rdmol/ProgressionSystem/pull/20/commits/d8183dbd485677952581c3a062b825b7faca4cb3